### PR TITLE
jwen/display friends list in profile

### PR DESCRIPTION
**Pull request by jiabaow**
display friends in recycler view after clicking on the Friends button in the profile.
show friends list by default once the user goes from search friends activity. 
rename friends to following.
Screen Shot 2022-05-17 at 21 39 02

 
all newly added code is tested.

### CODE REVIEW

File: app/src/androidTest/java/ch/sdp/vibester/database/PersistanceSetterTest.kt [hunk 1]
```diff
@@ -6,7 +6,7 @@ import org.junit.Test
 class PersistanceSetterTest {
     @Test
     fun testPersistance() {
-        PersistanceSetter.setPersistance()
-        assertEquals(true, PersistanceSetter.getPersistance())
+        PersistenceSetter.setPersistence()
+        assertEquals(true, PersistenceSetter.getPersistence())
```
Comment (MaximeZmt) [inline app/src/androidTest/java/ch/sdp/vibester/database/PersistanceSetterTest.kt:10]: Nice catch of the spelling mistake, you can also maybe rename the file for more coherence😅

File: app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt [hunk 2]
```diff
@@ -194,7 +194,7 @@ class QrScanningActivity : AppCompatActivity() {
                 camera.stop()
                 if (isTest || scannedValue in uidList){
                     if (!isTest) {
-                        usersRepo.setSubFieldValue(FireBaseAuthenticator().getCurrUID(),"friends", scannedValue, true)
+                        usersRepo.setSubFieldValue(FireBaseAuthenticator().getCurrUID(),"following", scannedValue, true)
```
Comment (MaximeZmt) [inline app/src/main/java/ch/sdp/vibester/activity/QrScanningActivity.kt:197]:
```suggestion
                        usersRepo.setSubFieldValue(FireBaseAuthenticator().getCurrUID(), "following", scannedValue, true)
```

File: app/src/main/java/ch/sdp/vibester/activity/profile/ProfileActivity.kt [hunk 3]
```diff
@@ -1,204 +1,226 @@
...
+        }
+    }
+
+    private fun showFriendsPosition(friends: MutableList<User>?) {
+        profileFriendsAdapter = ProfileFollowingAdapter(friends!!, this)
+        findViewById<RecyclerView>(R.id.profile_followingList)!!.adapter = profileFriendsAdapter
+    }
+
+
+    //check the UID here not sure
```
Comment (MaximeZmt) [inline app/src/main/java/ch/sdp/vibester/activity/profile/ProfileActivity.kt:87]: What is the purpose of this comment? is doubt resolve? Should we keep it ? 
If this is the case maybe add a // TODO

File: app/src/main/java/ch/sdp/vibester/activity/profile/ProfileActivity.kt [hunk 4]
```diff
@@ -1,204 +1,226 @@
...
+
+    private val imageSize = 1000
+    val imageRequestCode = 100
+
+    private var followings: MutableList<User> ? = null
+    private var profileFriendsAdapter: ProfileFollowingAdapter?= null
+
+    /**
+     * Generic onCreate method belonging to ProfileActivity.
+     */
```
Comment (MaximeZmt) [inline app/src/main/java/ch/sdp/vibester/activity/profile/ProfileActivity.kt:56]: not sure this is useful to redefine Javadoc for an override 😅

File: app/src/main/java/ch/sdp/vibester/activity/profile/ProfileActivity.kt [hunk 5]
```diff
@@ -1,204 +1,226 @@
...
+     * Generic listener for the return to main button.
+     */
+    private fun setRetToMainBtnListener() {
+        findViewById<FloatingActionButton>(R.id.profile_returnToMain).setOnClickListener {
+            IntentSwitcher.switch(this, MainActivity::class.java)
+            finish()
+        }
+    }
+
+    private fun setFollowingScoresBtnListener(btnId: Int, show: Int, hide: Int) {
```
Comment (MaximeZmt) [inline app/src/main/java/ch/sdp/vibester/activity/profile/ProfileActivity.kt:108]: For consistency reasons maybe add a Javadoc for the private function even if this is not necessary. But all the others in the same file have javadoc.

File: app/src/main/java/ch/sdp/vibester/activity/profile/ProfileActivity.kt [hunk 6]
```diff
@@ -1,204 +1,226 @@
...
+        }
+
+    }
+
+    private fun loadFollowing(followingMap: Map<String, Boolean>) {
+        followingMap.forEach { (userId, _) ->  dataGetter.getUserData(userId, this::addFollowing)}
+        showFriendsPosition(followings)
+    }
+
+    private fun addFollowing(following: User) {
```
Comment (MaximeZmt) [inline app/src/main/java/ch/sdp/vibester/activity/profile/ProfileActivity.kt:193]: same here? Javadoc?

File: app/src/main/java/ch/sdp/vibester/database/PersistenceSetter.kt [hunk 7]
```diff
@@ -1,33 +1,33 @@
...
+/**
+ * Enable Persistence in Firebase database.
+ * If internet connection is lost while app is on, it continues to work
+ */
+class PersistenceSetter private constructor(){
+    companion object{
+        private var alreadyExecuted : Boolean = false
+
+        /**
+         * Turn on the persistance
```
Comment (MaximeZmt) [inline app/src/main/java/ch/sdp/vibester/database/PersistenceSetter.kt:15]:
```suggestion
         * Turn on the persistence
```

File: app/src/main/java/ch/sdp/vibester/helper/ImageHelper.kt [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+import android.widget.ImageView
+import ch.sdp.vibester.api.BitmapGetterApi
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.async
+import kotlinx.coroutines.launch
+import java.util.concurrent.TimeUnit
+
+class ImageHelper {
+
```
Comment (MaximeZmt) [inline app/src/main/java/ch/sdp/vibester/helper/ImageHelper.kt:14]:
```suggestion
/**
* Missing Javadoc !!!
*/
class ImageHelper {

```

File: app/src/main/java/ch/sdp/vibester/helper/ImageHelper.kt [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    /**
+     * @param imageURI uri of the image
+     * @param avatar ImageView
+     * @param imageSize Int
+     * set the image of imageURI in avatar of given size
+     */
```
Comment (MaximeZmt) [inline app/src/main/java/ch/sdp/vibester/helper/ImageHelper.kt:20]:
```suggestion
    /**
     * Set the image of imageURI in avatar of given size
     * @param imageURI uri of the image
     * @param avatar ImageView
     * @param imageSize Int
     */
```

File: app/src/main/java/ch/sdp/vibester/user/ProfileFollowingAdapter.kt [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+import android.view.ViewGroup
+import android.widget.ImageView
+import android.widget.TextView
+import androidx.recyclerview.widget.RecyclerView
+import ch.sdp.vibester.R
+import ch.sdp.vibester.helper.AdapterHelper
+import ch.sdp.vibester.helper.loadImg
+
+/**
+ * set friends in userProfile with profile photo and username in RecycleView.
```
Comment (MaximeZmt) [inline app/src/main/java/ch/sdp/vibester/user/ProfileFollowingAdapter.kt:13]:
```suggestion
 * Set friends in userProfile with profile photo and username in RecycleView.
```

File: app/src/main/java/ch/sdp/vibester/user/ProfileFollowingAdapter.kt [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+        holder.bind(following[position])
+    }
+
+    override fun getItemCount() = following.size
+
+
+    inner class ProfileFollowingViewHolder(itemView: View): RecyclerView.ViewHolder(itemView), View.OnClickListener {
+
+        /**
+         * @param user with all the parameters
```
Comment (MaximeZmt) [inline app/src/main/java/ch/sdp/vibester/user/ProfileFollowingAdapter.kt:36]:
```suggestion
         * Purpose/Contexte of usage ? 
         * @param user with all the parameters
```

File: app/src/main/java/ch/sdp/vibester/user/UserScoreboardAdapter.kt [hunk 12]
```diff
@@ -10,11 +10,13 @@ import ch.sdp.vibester.R
 import ch.sdp.vibester.helper.AdapterHelper
 import ch.sdp.vibester.helper.loadImg
 
+/**
+ * set users list in recycleView in scoreboard
```
Comment (MaximeZmt) [inline app/src/main/java/ch/sdp/vibester/user/UserScoreboardAdapter.kt:14]:
```suggestion
 * Set users list in recycleView in scoreboard
```

File: app/src/main/res/values/strings.xml [hunk 13]
```diff
@@ -203,9 +202,15 @@
...
     <string name="PartyRoom_joinedPeople">Joined People:</string>
+
+    <!-- Profile -->
+    <string name="verify_change_profile_pic">Do you want to change your profile picture?</string>
     <string name="logout">logout</string>
     <string name="scores">Scores</string>
-    <string name="friends">Friends</string>
+    <string name="following">Following</string>
+    <string name="test_profile_image">bit.ly/3IUnyAF</string>
+    <string name="profile_username">Profile Username</string>
```
Comment (MaximeZmt) [inline app/src/main/res/values/strings.xml:213]: According to the [wiki](https://github.com/MaximeZmt/SDP_2022-Vibester/wiki) We now use the following convention for all new Strings: nameOfActivity_purposeOfString
Which means you should have 
- profile_verifyChangeProfilePic
- profile_logout
- profile_scores
- ...

In addition you have to put them between already cleaned string and the old string. Meaning around line 42 of string.xml

**Review by MaximeZmt — COMMENTED** [review]
Nice Work 😀👍 Left a list of comments below!

File: app/src/main/java/ch/sdp/vibester/activity/profile/ProfileActivity.kt (hunk 3 above)
Comment (jiabaow) [reply app/src/main/java/ch/sdp/vibester/activity/profile/ProfileActivity.kt:87]: I don't know... I didn't write this part.

File: app/src/main/java/ch/sdp/vibester/activity/profile/ProfileActivity.kt (hunk 4 above)
Comment (jiabaow) [reply app/src/main/java/ch/sdp/vibester/activity/profile/ProfileActivity.kt:56]: I didn't add it but I can remove it.

File: app/src/main/java/ch/sdp/vibester/activity/profile/MyProfileActivity.kt [hunk 14]
```diff
@@ -53,13 +54,20 @@ class MyProfileActivity : ProfileActivity() {
...
+            showDialogWhenChangeImage()
         }
         findViewById<CardView>(R.id.myCardView).setOnClickListener {
-            showGeneralDialog( "Do you want to change your profile picture?", false)
+            showDialogWhenChangeImage()
         }
     }
 
+    /**
+     * helper function to de-duplicate
```
Comment (kamilababayeva) [inline app/src/main/java/ch/sdp/vibester/activity/profile/MyProfileActivity.kt:65]:
```suggestion
     * helper function to show dialog 
```

File: app/src/main/java/ch/sdp/vibester/activity/profile/MyProfileActivity.kt [hunk 15]
```diff
@@ -53,13 +54,20 @@ class MyProfileActivity : ProfileActivity() {
      */
     private fun setChangeImageBtnListener() {
         findViewById<ImageView>(R.id.avatar).setOnClickListener {
-            showGeneralDialog( "Do you want to change your profile picture?", false)
+            showDialogWhenChangeImage()
         }
         findViewById<CardView>(R.id.myCardView).setOnClickListener {
-            showGeneralDialog( "Do you want to change your profile picture?", false)
+            showDialogWhenChangeImage()
```
Comment (kamilababayeva) [inline app/src/main/java/ch/sdp/vibester/activity/profile/MyProfileActivity.kt:60]: do you know why there are two clicks for the same functionality?

**Review by kamilababayeva — APPROVED** [review]
Thanks for your changes! looks good to me!

File: app/src/main/java/ch/sdp/vibester/activity/profile/MyProfileActivity.kt (hunk 15 above)
Comment (jiabaow) [reply app/src/main/java/ch/sdp/vibester/activity/profile/MyProfileActivity.kt:60]: no, I just saw the duplication and refactored them.

Comment (jiabaow) [reply app/src/main/java/ch/sdp/vibester/activity/profile/MyProfileActivity.kt:60]: > no, I just saw the duplication and refactored them.

I will ask in the group

File: app/src/main/java/ch/sdp/vibester/activity/profile/ProfileActivity.kt (hunk 6 above)
Comment (MaximeZmt) [reply app/src/main/java/ch/sdp/vibester/activity/profile/ProfileActivity.kt:195]: I think you forget these two 😅

File: app/src/main/java/ch/sdp/vibester/database/PersistenceSetter.kt (hunk 7 above)
Comment (MaximeZmt) [reply app/src/main/java/ch/sdp/vibester/database/PersistenceSetter.kt:15]: and that one too

File: app/src/main/java/ch/sdp/vibester/helper/ImageHelper.kt (hunk 8 above)
Comment (MaximeZmt) [reply app/src/main/java/ch/sdp/vibester/helper/ImageHelper.kt:14]: and that one

Comment (MaximeZmt) [reply app/src/main/java/ch/sdp/vibester/helper/ImageHelper.kt:14]: oki, dunno what's the best, let's say this is fine for now 👍

**Review by MaximeZmt — APPROVED** [review]
LGTM 👍

**Pull request merged**